Docker Container및 Image시각화 구현 docker Container및 Image시각화 구현

kimykstar opened the pull request:
## 작업 개요

Resolves #28 Resolves #36 Resolves #44 Resolves #52 Resolves #68 Resolves #84 Resolves #60 Resolves #76 

## 작업 상세 내용

Container와 Image를 시각화 하기 위한 기능들을 추가하였습니다.
- Container가 어떤 Image로부터 나왔는지 알기 위한 색상 동기화
- Container의 생성(created), 실행(start, run), 삭제(rm), 상태 변경(stop 등)들의 애니메이션 동작 및 시각화 구현

## 참고자료(선택)

[Tailwind Animation](https://tailwindcss.com/docs/animation)
[개발 일지](https://github.com/boostcampwm-2024/web34-LearnDocker/wiki/%5B4%EC%A3%BC-4%EC%9D%BC%EC%B0%A8-%E2%80%90-J048-%EA%B9%80%EC%98%81%EA%B4%80%5D-%EA%B0%9C%EB%B0%9C-%EC%9D%BC%EC%A7%80%28%ED%94%84%EB%A1%A0%ED%8A%B8-%EB%AC%B8%EC%A0%9C-%ED%95%B4%EA%B2%B0-%EB%B0%8F-%EC%8B%9C%EA%B0%81%ED%99%94-%EC%9E%91%EC%97%85-%EC%B4%88%EA%B8%B0%EC%84%A4%EC%A0%95%29)

## 문제점 혹은 고민(선택)
- 시각화를 위한 함수 로직이 변화된 image의 길이, container의 길이를 이용해 판단하다 보니, 코드가 생각보다 너무 복잡해졌습니다.. 일단 기능은 이대로 가져가고, 나중에 리팩토링 시간에 조금 더 개선하는 것이 좋아보입니다.
- 호스트 컨테이너에서 image와 container의 상태가 완전히 동기화가 되는 상황이 아닙니다. 
    - 호스트 컨테이너에서 container가 잘 동작하다가 중간에 예키치 못하게 중단되더라도, 사용자가 docker명령을 입력하지 않으면, 호스트 컨테이너의 상태를 조회하는 요청이 가지 않기 때문에 완전한 동기화가 되지 않습니다.(polling 혹은 sse를 통해 나중에 리팩토링이 필요해보입니다.)

kimykstar committed:
fix: docker pull 시 이미지 색상 겹침 문제 해결

- docker pull명령으로 2개 이상의 이미지를 받았을 때 색상이 다르게
  입혀지는게 아니고, 같은 색상으로 입혀지는 문제발생

kimykstar committed:
feat: container애니메이션 추가

- Container애니메이션 type, Arrow 추가
- Container 생성 및 삭제에 대한 애니메이션 추가

kimykstar committed:
feat: container 시각화 애니메이션 구현

- container애니메이션 화살표 위치 변경
- container색상 변경 기능 추가

kimykstar committed:
feat: 컨테이너 상태에 따른 애니메이션 추가

- tailwind ping애니메이션 설정(tailwind.config.js)
- container시각화에만 애니메이션이 적용되도록 수정
- console.log지우기

kimykstar committed:
feat: docker run 애니메이션 추가 및 구현

- docker run명령에 대한 애니메이션 추가
- docker run명령에 대한 State갱신
- State갱신에 따른 시각화 변경

kimykstar committed:
chore: 안쓰는 함수 제거 및 console.log제거

kimykstar committed:
feat: Container상태 추가

- container상태에 따른 색상 추가(tailwind.config.js)
- Docker Container상태 상수화

kimykstar committed:
feat: docker 상태 변경 명령어 시각화 추가

- 컨테이너 상태 변경 애니메이션 추가
- docker stop명령어 상수화
- docker stop명령어 입력에 따른 동작 추가

kimykstar committed:
feat: 상태 변경 명령어에 대해서만 애니메이션 적용

- 상태 변경 명령어가 아닌 docker ps, docker images의 경우에도
  애니메이션이 적용됨
- 상태 변경 명령어에만 애니메이션이 적용되도록 수정

kimykstar committed:
feat: Container상태 색상 추가

- Container상태에 따른 색상 추가
- created, paused, stopped, dead상태

tuchongkim commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -15,28 +15,39 @@ const useDockerVisualization = () => {
         isVisible: false,
         key: 0,
     });
+    const STATE_CHANGE_COMMAND_REGEX =
+        /^docker\s+(run|create|start|stop|pull|rmi|rm|restart|pause|unpause|rename|attach|tag|build|load|commit|kill)(\s|$)/;
+    const colors = ['#000000', '#FFC107', '#4CAF50', '#2196F3', '#673AB7', '#E91E63'];
+
+    const getNotUsedColor = (images: Image[]) => {
+        const notUsedColors = colors.filter((color) => {
+            return !images.some((image) => image.color === color);
+        });
+        return notUsedColors[0];
+    };
```
상태 변화가 일어나는 docker 명령어들을 잘 정리해주셨네요. 사용되지 않은 색상 적용도 잘 해주신 것 같습니다!

guune commented:
> 호스트 컨테이너에서 image와 container의 상태가 완전히 동기화가 되는 상황이 아닙니다.

이거는 어쩔 수 없는 것 같습니다. 엔터를 칠때 마다 확인을 하니 이 정도면 괜찮을 것 같아요

guune commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -81,15 +87,73 @@ const useDockerVisualization = () => {
+        setContainers((currentContainers) => {
+            if (currentContainers.length === newContainers.length) {
+                if (isChangedContainerStatus(currentContainers, newContainers)) {
+                    setDockerOperation(DOCKER_OPERATIONS.CONTAINER_STOP);
```
container가 stop이였다가 start 되는 경우는 아직 없는 걸까요?

guune replied:
코드를 다시 보니깐 상태 변화가 생기면 컨테이너 노드에서 알아서 처리가 될 것 같습니다.
맞을까요? 맞다면 CONTAINER_STOP 뿐만 아니라 CONTAINER 상태를 모두 포괄하는 operation 이름이 좋아 보여요

kimykstar replied:
네 맞습니다. 컨테이너 상태가 변화되면 알아서 처리되도록 하였습니다. CONTAINER_STATUS_CHANGED로 수정했습니다.

guune commented on frontend/src/components/visualization/ArrowAnimation.tsx:
```diff
@@ -60,6 +60,129 @@ export const ArrowAnimation = ({ isVisible, onComplete, dockerOperation }: Anima
+                        isVisible={isVisible}
+                        onAnimationEnd={onComplete}
+                    />
+                    <Arrow
```
이미지가 없는 경우 아예 화살표 애니매이션을 합치셨군요 좋습니다.

guune commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -81,15 +87,73 @@ const useDockerVisualization = () => {
+                    setDockerOperation(DOCKER_OPERATIONS.CONTAINER_STOP);
+                    const elements = setColorToElements(newImages, newContainers);
+                    setPendingContainers(elements.initContainers);
+                    if (command.match(STATE_CHANGE_COMMAND_REGEX))
```
위에서 isChangedContainerStatus으로 컨테이너 상태가 변했는지 확인했는데 커맨드를 다시 한 번 확인 하시는 이유가 있을까요?

아 이 경우는 컨테이너가 사용자의 명령어 말고 컨테이너가 스스로 죽었을때 화살표 애니매이션을 안 나오게 하는 장치일까요?

kimykstar replied:
컨테이너가 스스로 죽고 사용자는 docker ps 혹은 docker images명령을 치게 되면, api요청을 통해 image와 contianer목록을 가져오게 되는데 이 때 상태가 변경되었음이 판단이 되어 애니메이션이 동작합니다. 따라서, 상태 변화 명령어가 들어올 때만 애니메이션이 적용될 수 있도록 수정한 것입니다.

guune commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -81,15 +87,73 @@ const useDockerVisualization = () => {
     };
 
-    const updateVisualizationData = async () => {
+    const isChangedContainerStatus = (
+        prevContainers: Container[],
+        currentContainers: Container[]
+    ) => {
+        const result = prevContainers.reduce<Container[]>((reducer, prevContainer) => {
+            const matchedContainer = currentContainers.find((currentContainer) => {
+                return currentContainer.id === prevContainer.id;
+            });
+            if (!matchedContainer) {
+                throw new Error('isChangedContainerStatue함수 undefind 에러');
+            }
+            if (matchedContainer.status !== prevContainer.status) {
+                reducer.push(matchedContainer);
+            }
+            return reducer;
+        }, []);
+        return result.length > 0 ? true : false;
+    };
+
```
```suggestion
const isChangedContainerStatus = (
    prevContainers: Container[],
    currentContainers: Container[]
) => {
    return prevContainers.some(prevContainer => {
        const matchedContainer = currentContainers.find(
            currentContainer => currentContainer.id === prevContainer.id
        );
        if (!matchedContainer) {
            throw new Error('isChangedContainerStatus함수 undefined 에러');
        }
        return matchedContainer.status !== prevContainer.status;
    });
};
```

reduce 말고 some 고차함수도 괜찮아 보입니다.

kimykstar committed:
Update frontend/src/hooks/useDockerVisualization.ts

Co-authored-by: guune <108771739+guune@users.noreply.github.com>

kimykstar committed:
refactor: DOCKER_OPERATION명 수정

- DOCKER_OPERATIONS.STOP을 DOCKER_OPERATIONS.CONTAINER_STATUS_CHANGED로
  수정
- 상수 변경에 따른 애니메이션 CASE문 명령명 변경

SH9480P commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -15,28 +15,39 @@ const useDockerVisualization = () => {
 
     const setColorToElements = (images: Image[], containers: Container[]) => {
-        const initImages = images.map((image, index) => {
+        const sortedImages = images.sort((a, b) => a.id.localeCompare(b.id));
```
sortFunction을 인자로 넘겨주지 않아도 원하는 결과를 얻을 수 있을 것 같아요!
그런데 왜 여기서 정렬을 수행해주나요?

kimykstar replied:
아 정렬을 수행하지 않으면 이미지와 컨테이너 색상이 동일하게 입혀지지 않는 문제가 발생했습니다.. 그래서 수정을 했던 건데 지금 보니 아직도 오류가 있네요 정렬은 현재 필요하지 않은 것 같고 다시 수정해봐야겠습니다.

SH9480P commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -56,15 +67,10 @@ const useDockerVisualization = () => {
+        const { images: newImages, containers: newContainers } = data;
         setImages((currentImages) => {
             if (currentImages.length === newImages.length) {
-                setAnimation((prev) => ({
-                    isVisible: false,
-                    key: prev.key,
-                }));
```
이 부분은 불필요해서 없애신 게 맞나요?

kimykstar replied:
맞습니다. 이미지에 변경사항이 없을 때 애니메이션을 끄는 상태변화 함수가 필요한가? 생각이 들었습니다. 이미지의 경우 컨테이너처럼 상태에 대해서 시각화를 고려해야 할 상황이 아니기 때문에 길이가 동일하다면, 바로 setImages내 콜백 함수를 종료시키면 되겠다 생각했습니다.

SH9480P commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -81,15 +87,68 @@ const useDockerVisualization = () => {
+                            isVisible: true,
+                            key: prev.key + 1,
+                        }));
+                    else setContainers(elements.initContainers);
```
setContainer 내부에 있는 코드이니만큼 `return elements.initContainers`를 해도 되지 않을까 싶은데..
`return elements.initContainers`를 하면 문제가 되나요?

SH9480P commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -81,15 +87,68 @@ const useDockerVisualization = () => {
+            if (currentContainers.length === newContainers.length) {
+                if (isChangedContainerStatus(currentContainers, newContainers)) {
+                    setDockerOperation(DOCKER_OPERATIONS.CONTAINER_STATUS_CHANGED);
+                    const elements = setColorToElements(newImages, newContainers);
```
여기서 setColorToElements가 필요한 이유가 뭔가요?
container의 추가 및 삭제가 없고, 상태만 변경된 거라면 이미지와 container의 색상도 달라질 일이 없지 않나요?

kimykstar replied:
위 코드에서 사용하는 newContainers에는 color가 없는 상태입니다. 따라서, 응답받은 Image들의 색상을 입히고, 그 색상들을 기반으로 Container의 색상을 초기화 해주는 과정이 필요합니다!

SH9480P commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -98,12 +157,30 @@ const useDockerVisualization = () => {
 
     const handleAnimationComplete = () => {
-        setImages(pendingImages);
+        if (
+            dockerOperation === DOCKER_OPERATIONS.IMAGE_PULL ||
+            dockerOperation === DOCKER_OPERATIONS.IMAGE_DELETE
+        )
+            setImages(pendingImages);
+        if (
+            dockerOperation === DOCKER_OPERATIONS.CONTAINER_CREATE ||
+            dockerOperation === DOCKER_OPERATIONS.CONTAINER_DELETE ||
+            dockerOperation === DOCKER_OPERATIONS.CONTAINER_STATUS_CHANGED
+        ) {
+            setContainers(pendingContainers);
+        }
+        if (dockerOperation === DOCKER_OPERATIONS.CONTAINER_RUN) {
+            setImages(() => {
+                setTimeout(() => {
+                    setContainers(pendingContainers);
+                }, 0);
+                return pendingImages;
+            });
+        }
```
- docker run으로 컨테이너를 다운로드받고 실행시킨 상황에서 setContainers가 setImages 안에 들어가야하는 이유가 뭔가요?
- setContainers는 왜 setTimeout 0으로 지연 처리되어야 하나요?

kimykstar replied:
아 이 코드 이상하게 Container만 갱신이 되고, Image는 갱신이 안되는 상황이 발생했습니다.. 근데 방금 
```
setImages(pendingImages);
setContainers(pendingContainers);
```
위 코드로 변경해도 잘 동작 하네요! 수정하겠습니다.

SH9480P commented on frontend/src/components/visualization/ContainerNode.tsx:
```diff
@@ -20,15 +39,25 @@ const ContainerNode = ({
+                        key={element.id}
                     >
-                        {name}
+                        {isContainer(element) && (
+                            <span
+                                className={`absolute -top-1 -right-1  inline-flex h-2 w-2 rounded-full ${STATUS_COLORS[element.status]} opacity-75 animate-ping`}
+                            ></span>
+                        )}
+                        {isContainer(element) && (
+                            <span
+                                className={`absolute -top-1 -right-1 inline-flex rounded-full h-2 w-2 ${STATUS_COLORS[element.status]}`}
+                            ></span>
+                        )}
```
```suggestion
                        {isContainer(element) && (
                            <>
                                <span
                                    className={`absolute -top-1 -right-1  inline-flex h-2 w-2 rounded-full ${STATUS_COLORS[element.status]} opacity-75 animate-ping`}
                                ></span>
                                <span
                                    className={`absolute -top-1 -right-1 inline-flex rounded-full h-2 w-2 ${STATUS_COLORS[element.status]}`}
                                ></span>
                            </>
                        )}
```

SH9480P reviewed (commented):
리뷰가 늦어서 죄송합니다.
개인적으로 코드가 잘 이해되지 않네요.. 😢
그래서 코드의 의도에 대한 질문이 많은데요, 설명하기 어려우시면 허들이든 zep이든 직접 얘기해주셔도 좋아요
고생하셨습니다!

kimykstar committed:
Update frontend/src/components/visualization/ContainerNode.tsx

Co-authored-by: Park Se Hwan <64393421+SH9480P@users.noreply.github.com>

tuchongkim commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -15,28 +15,39 @@ const useDockerVisualization = () => {
     };
 
     const setColorToElements = (images: Image[], containers: Container[]) => {
-        const initImages = images.map((image, index) => {
+        const sortedImages = images.sort((a, b) => a.id.localeCompare(b.id));
```
혹시 이 setColorToElements() 메서드를 어떤 상황에서 어떤 용도로 사용하는지 알 수 있을까요?

kimykstar replied:
image와 Container의 색상을 한번에 입혀줄 때 사용하고 있습니다. 사실 서버에 요청 후 응답받는 데이터들이 color가 없기 때문에 image나 container의 상태가 변화할 때 자주 사용하고 있습니다.

kimykstar committed:
fix: Image, Container색상 맞추기

- Container가 어떤 Image로부터 생성되었는지 확인하기 위해 색상 동일화
  하기
- Index를 기반으로 색상 지정이 아닌 사용되지 않는 색상으로 지정될 수
  있도록 동작 개선

kimykstar committed:
fix: 빈 문자열의 결과응답 시 처리 오류 수정

- alpine 및 ubuntu의 이미지를 run했을 때 응답 문자열은 빈 문자열이었기
  때문에 명령어에 따른 시각화가 되지 않은 문제 수정
- container색상 입히기 로직 수정

kimykstar committed:
refactor: handleTerminalEnterCallback리팩토링

- setImages, setContainer콜백함수를 활용해 갱신하는 것이 아닌 useRef훅을
  활용해 최신 정보를 유지 및 갱신하도록 동작 개선

kimykstar committed:
Merge branch 'feature-#28-container-visualization' of https://github.com/boostcampwm-2024/web34-LearnDocker into feature-#28-container-visualization

tuchongkim commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -15,95 +15,197 @@ const useDockerVisualization = () => {
-        return newImages.map((newImage, index) => {
+        return newImages.map((newImage) => {
             const prevImage = prevImages.find((img) => img.id === newImage.id);
-            if (prevImage) {
+            if (prevImage && Object.keys(prevImage).includes('color')) {
                 return prevImage;
             }
-            return {
+            const newData = {
                 ...newImage,
-                color: colors[index % colors.length],
+                color: getNotUsedColor(prevImages),
             };
+            prevImages.push(newData);
+            return newData;
         });
```
color 속성이 있는 경우와 없는 경우를 나눠주신 것 아주 좋습니다 ㅎㅎ

tuchongkim commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -15,95 +15,197 @@ const useDockerVisualization = () => {
+        return notUsedColors[0];
+    };
+
+    const setColorToElements = (prevImages: Image[], images: Image[], containers: Container[]) => {
+        const initImages = updateImageColors(images, prevImages);
+        const initContainers = updateContainerColors(initImages, containers);
```
😊

guune commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -15,95 +15,197 @@ const useDockerVisualization = () => {
+    const STATE_CHANGE_COMMAND_REGEX =
+        /^docker\s+(run|create|start|stop|pull|rmi|rm|restart|pause|unpause|rename|attach|tag|build|load|commit|kill)(\s|$)/;
+    const imagesRef = useRef<Image[]>();
+    const containersRef = useRef<Container[]>();
```
useRef 사용하셔서 리팩토링 하셨네요 좋습니다 👍

guune commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -15,95 +15,197 @@ const useDockerVisualization = () => {
+        const { images: newImages, containers: newContainers } = data;
+        const prevImages = imagesRef.current as Image[];
+        const prevContainers = containersRef.current as Container[];
+        console.log(command);
```
삭제해도 될 것 같습니다.

kimykstar replied:
바로 삭제해버렸습니다 ㅎㅎ

guune commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -15,95 +15,197 @@ const useDockerVisualization = () => {
+                    isVisible: true,
+                    key: prev.key + 1,
+                }));
+            else setContainers(updatedContainers);
```
이 긴 if , else if , else를 나눌 수 있을 것 같은데 딱 떠오르는 좋은 방법은 없네요 ㅜ

kimykstar replied:
저도 이 부분 너무 생각이 안나서 일단은 조건문 다른 모듈로 뺐습니다 ㅎㅎㅎ 뺀 모듈에서 리턴받은 값을 기반으로 switch case문으로 동작하도록 최종 결정했습니다.

guune commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -15,95 +15,197 @@ const useDockerVisualization = () => {
 
     const handleAnimationComplete = () => {
-        setImages(pendingImages);
+        if (
+            dockerOperation === DOCKER_OPERATIONS.IMAGE_PULL ||
+            dockerOperation === DOCKER_OPERATIONS.IMAGE_DELETE
+        )
+            setImages(pendingImages);
+        if (
+            dockerOperation === DOCKER_OPERATIONS.CONTAINER_CREATE ||
+            dockerOperation === DOCKER_OPERATIONS.CONTAINER_DELETE ||
+            dockerOperation === DOCKER_OPERATIONS.CONTAINER_STATUS_CHANGED
+        ) {
+            setContainers(pendingContainers);
+        }
+        if (dockerOperation === DOCKER_OPERATIONS.CONTAINER_RUN) {
+            setImages(pendingImages);
+            setContainers(pendingContainers);
+        }
```
>사용자가 제공한 새로운 값이 [Object.is](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Object/is)에 의해 현재 state와 동일하다고 판정되면, React는 컴포넌트와 그 자식들을 리렌더링하지 않습니다.

https://ko.react.dev/reference/react/useState

그냥 조건문 다 없애고 
setImages(pendingImages);
setContainers(pendingContainers);
이것만 있어도 괜찮을것 같은데 어떻게 생각하시나요

kimykstar replied:
맞습니다! 결국 useRef를 활용해서 최신 상태의 container와 image데이터들을 가지고 있는 상태라서 pendingImages와 pendingContainers 모두 지웠습니다!

kimykstar committed:
refactor: 상수, 함수, 모듈 분리

- COLORS 및 Container상태 변경 명령 정규표현식
  constant/visualization.ts로 분리
- hook을 포함하지 않는 함수들 utils/visualizationUtils.ts로 분리
- Image, Container State Visualization타입으로 통일하여 관리
- 함수의 매개변수 Visualization타입으로 통일
- handleTerminalEnterCallback함수 switch case문으로 수정

kimykstar committed:
chore: 불필요한 export 제거

tuchongkim commented on frontend/src/hooks/useDockerVisualization.ts:
```diff
@@ -1,122 +1,149 @@
+        const newElements = data;
+        const operation = getDockerOperation(prevElements, newElements);
+
+        switch (operation) {
+            case DOCKER_OPERATIONS.IMAGE_PULL:
+                handleImageStateLengthChange(prevElements, newElements, operation);
+                break;
+            case DOCKER_OPERATIONS.IMAGE_DELETE:
+                handleImageStateLengthChange(prevElements, newElements, operation);
+                break;
+            case DOCKER_OPERATIONS.CONTAINER_CREATE:
+                handleContainerStateLengthChange(newElements, operation);
+                break;
+            case DOCKER_OPERATIONS.CONTAINER_RUN:
+                handleElementsStateLengthChange(newElements);
+                break;
+            case DOCKER_OPERATIONS.CONTAINER_DELETE:
+                handleContainerStateLengthChange(newElements, operation);
+                break;
+            case DOCKER_OPERATIONS.CONTAINER_STATUS_CHANGED:
+                handleContainerStateChange(newElements, command);
+                break;
```
switch case 문으로 알아보기 쉽게 잘 작성해주신 것 같습니다!

guune commented:
긴 시각화 과정 고생하셨습니다. 커밋을 작업 중간중간 해주셔서 리팩토링을 어떻게 하셨는지도 쉽게 파악할 수 있어서 좋았습니다. 리팩토링 된 코드는 깔끔하게 잘 짜신 것 같습니다.
